scipy/sparse/csgraph/_flow.pyx

```diff
@@ -27,7 +27,7 @@ class MaximumFlowResult:
         return 'MaximumFlowResult with value of %d' % self.flow_value 
 
 
-def maximum_flow(csgraph, source, sink):
+def maximum_flow(csgraph, source, sink, *, method='edmonds_karp'):
```

Comment:
If we expect the new method to almost always be better, and if the rest of the API is left the same, I don't see why we wouldn't just let `dinic` be the default value of `method` here.

Reply:
This might change user's downstream application code for some problem as both algorithm have different complexity.

For V the set of vertices and E the set of edges:
 - Dinic's is O(|V|² ⨉ |E|)
 - Edmonds–Karp's is O(|V| ⨉ |E|²)

Reply:
But is that a problem? |_V_| < |_E_|, so looking only at complexity, this is always an improvement.

I'm sure there are examples where the constants play out in Edmonds-Karp's favor; if the density is sufficiently low, or for particular graph structures, but I'm not immediately convinced that those are not so special that the improved UX from having a faster default solver is better.

Reply:
That's true. There are configuration where Edmonds-Karp's performs better but they are rare.

Hence, I would also make Dinic's the default.

Reply:
So, should I make Dinic's algorithm default?

Reply:
We cannot do that, it would break backward compatibility. For this to happen, we would need a deprecation cycle.

Reply:
Okay. May be we can wait till next version to do that. Like if users try out `dinic` and they like it then we can make it default after a few versions. Let `edmonds-karp` be the default for now.

Reply:
What you could do right now though, is to add a note describing a bit more the differences between these algorithms (the time complexity thing for instance is nice) and why `dinic` would be a sensible choice. There is already some text about `edmonds_karp` in that regard.

Reply:
Why would that break backwards compatibility, and what would be the problem in the algorithm changing? The addition of a parameter is API stable, and I don't see a compelling reason to default to an implementation that's much slower in general.

Reply:
You are the expert here so please correct me if I am wrong in the following 😃

The API is stable yes, but the algorithm is different. Meaning that some users might expect to have `edmonds-karp`, just this is a breaking change.

What about the different complexity, for some problems this might have a negative impact no?

Something else we did not discuss above is that this new implementation is new. Meaning we could have unknown/corner cases bugs. Hence again, break things.

Reply:
You are the expert here so please correct me if I am wrong in the following 😃

The API is stable yes, but the algorithm is different. Meaning that some users might expect to have `edmonds-karp`, just this is a breaking change.

What about the different complexity, for some problems this might have a negative impact no?

Something else we did not discuss above is that this new implementation is new. Meaning we could have unknown/corner cases bugs. Hence again, break things.

Reply:
I will go with keeping 'edmonds-karp' as default for now. After user feedback (if they try out 'dinic' and they like it) then we can make it default in future versions.

Reply:
@tupui The complexity itself would never have negative impact, since |_V_| ≤ |_E_| + 1 (the extreme case being one where the graph is a tree), so _O_(|_V_|² ⨉ |_E_|) is a subset of _O_(|_V_| ⨉ |_E_|²). The constants might favor Edmonds--Karp in some situations, and you can probably conjure an example where the old one is faster, but chances are they're rare (note that Dinic's algorithm is also based on shortest augmenting paths); if I get the chance to test before this gets in, I'll see if I can come up with some examples as well.

All in all, I think the improved UX from having a fast default is preferable.

Another way of looking at what can happen by changing vs. not doing it:

* Changing: The default case is that client code now becomes faster; good times. In the (presumably rare) case that code is negatively impacted, chances are users will check to see what happened, find the new `method` parameter, and immediately get back to working code.
* Not changing: The default case is code remaining slow, and chances are that the user, unless they're keeping a very close eye on the release notes, will not notice that a new `method` parameter has arrived that they would likely be able to make use of.

Reply:
Thank you @fuglede for the discussion. I am bought so removing my reserve. Let see if others drop-by.

Reply:
Sure. Anyways, the interface is still the same so it might not be a breaking change IMHO. It's just that the running time will now differ.

Reply:
Done. Let me know if I should make 'dinic' as the default.

Reply:
I'm +1 on changing the default to `'dinic'`. It's backwards compatible, the lower algorithmic complexity is compelling, and the change doesn't seem particularly high-risk for introducing bugs/regressions.

There's a couple of similar cases in SciPy that come to mind:
- `scipy.cluster` had some algorithms with higher complexity than `fastcluster`. The algorithms were rewritten from scratch, and the performance difference was large enough that we threw away the old implementation rather than have a `method` keyword.
- `optimize.linprog` now has faster non-default methods (the `'highs-xxx'` ones). Those were added as non-default, the plan is to switch in the future to them with warnings. The switch could not be made straight away because the algorithms have heuristics and may therefore introduce new failures on problems that were solvable before. So just switching over was too backwards incompatible.

This case is more like the `cluster` one, except that we also keep `'edmonds-karp'` because it's a well-known algorithm that still has value to keep.

Reply:
Okay. I will change the default to `'dinic'`. Thanks for the discussions. Let's see how it goes.

Reply:
Do we want to track the benchmarks of the methods separately, i.e. also update `benchmarks/benchmarks/sparse_csgraph_maxflow.py`?

Reply:
Do you mean to check for `edmonds_karp` as well? Because `dinic` is already being bench marked by default by the current code in the file you referenced. Let me know.

Reply:
Yep, that's what I was wondering; to make sure we catch regressions for `edmonds_karp` (or any by-method performance regressions). (Does SciPy do so elsewhere?)